frontend/src/concepts/pipelines/topology/parseUtils.ts

```diff
@@ -159,26 +161,133 @@ export const parseRuntimeInfo = (
+  }
+
+  const execution = executions.find(
+    (e) => e.getCustomPropertiesMap().get('task_name')?.getStringValue() === taskId,
+  );
+
+  if (!execution) {
+    return undefined;
+  }
+
+  const lastUpdatedTime = execution.getLastUpdateTimeSinceEpoch();
+  let completeTime;
+  const lastKnownState = execution.getLastKnownState();
+  // Logic comes from https://github.com/opendatahub-io/data-science-pipelines/blob/master/frontend/src/components/tabs/RuntimeNodeDetailsV2.tsx#L245-L253
+  if (
+    lastUpdatedTime &&
+    (lastKnownState === Execution.State.COMPLETE ||
+      lastKnownState === Execution.State.FAILED ||
+      lastKnownState === Execution.State.CACHED ||
+      lastKnownState === Execution.State.CANCELED)
+  ) {
+    completeTime = new Date(lastUpdatedTime).toISOString();
+  }
+  return {
+    startTime: new Date(execution.getCreateTimeSinceEpoch()).toISOString(),
+    completeTime,
+    state: getResourceStateText({
+      resourceType: ResourceType.EXECUTION,
+      resource: execution,
+    }),
+    taskId: `task.${taskId}`,
+    podName: execution.getCustomPropertiesMap().get('pod_name')?.getStringValue(),
+  };
+};
+
+export enum ResourceType {
+  ARTIFACT = 'ARTIFACT',
+  EXECUTION = 'EXECUTION',
+}
+
+export interface ArtifactProps {
+  resourceType: ResourceType.ARTIFACT;
+  resource: Artifact;
+}
+
+export interface ExecutionProps {
+  resourceType: ResourceType.EXECUTION;
+  resource: Execution;
+}
+
+export type ResourceInfoProps = ArtifactProps | ExecutionProps;
+
+// Get text representation of resource state.
+// Works for both artifact and execution.
+export const getResourceStateText = (
+  props: ResourceInfoProps,
+): ArtifactStateKF | ExecutionStateKF | undefined => {
+  if (props.resourceType === ResourceType.ARTIFACT) {
+    const state = props.resource.getState();
+    switch (state) {
+      case Artifact.State.PENDING:
+        return ArtifactStateKF.PENDING;
+      case Artifact.State.LIVE:
+        return ArtifactStateKF.LIVE;
+      case Artifact.State.MARKED_FOR_DELETION:
+        return ArtifactStateKF.MARKED_FOR_DELETION;
+      case Artifact.State.DELETED:
+        return ArtifactStateKF.DELETED;
+      default:
+        return undefined;
+    }
+  } else {
+    // type == EXECUTION
+    const state = props.resource.getLastKnownState();
+    switch (state) {
+      case Execution.State.NEW:
+        return ExecutionStateKF.NEW;
+      case Execution.State.RUNNING:
+        return ExecutionStateKF.RUNNING;
+      case Execution.State.COMPLETE:
+        return ExecutionStateKF.COMPLETE;
+      case Execution.State.CANCELED:
+        return ExecutionStateKF.CANCELED;
+      case Execution.State.FAILED:
+        return ExecutionStateKF.FAILED;
+      case Execution.State.CACHED:
+        return ExecutionStateKF.CACHED;
+      default:
+        return undefined;
+    }
+  }
+};
+
+export const translateStatusForNode = (
+  state?: RuntimeStateKF | ExecutionStateKF | ArtifactStateKF,
```

Comment:
nit: I think a more comprehensive approach would be to have separate switch statements for the different types of state being used here.

Reply:
That would be complex because we need to change the type of the state to something like `{type: xxx, state: xxx}` and I checked the state is used across a few files, which need some refactors to the use of these files, and I don't think it worth the effort just to try to separate the types. Basically, we are still doing the switch statement based on the string value.

Reply:
Why wouldn't there just be 3 functions, 1 for artifacts, 1 for executions and 1 for runtime?

Reply:
For example, if you separate this into 3 functions, which function should be called here? https://github.com/opendatahub-io/odh-dashboard/blob/main/frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/SelectedNodeDetailsTab.tsx#L39

Reply:
Theoretically there is something there that you could use to determine which function to use, and you'd conditionally use them based on whether you're dealing with artifact, experiment, or runtime state. If that's not possible then my idea would not work.

Reply:
Yes I wanted, but it's not something that could be easily implemented for now, because the `RuntimeState` is different. We always try to use MLMD first, and when it's not available, we parse the runtime state as a backup plan instead of showing users an error state. So it's hard to separate them.